lambdas/s3hash/src/t4_lambda_s3hash/__init__.py

```diff
@@ -0,0 +1,374 @@
+from __future__ import annotations
+
+import asyncio
+import base64
+import contextlib
+import contextvars
+import functools
+import hashlib
+import logging
+import math
+import os
+import typing as T
+
+import aiobotocore.config
+import aiobotocore.response
+import aiobotocore.session
+import botocore.exceptions
+import pydantic
+import tenacity
+
+from quilt_shared.aws import AWSCredentials
+from quilt_shared.pkgpush import Checksum as ChecksumBase
+from quilt_shared.pkgpush import (
+    ChecksumResult,
+    ChecksumType,
+    MPURef,
+    S3ObjectDestination,
+    S3ObjectSource,
+)
+
+if T.TYPE_CHECKING:
+    from types_aiobotocore_s3.client import S3Client
+
+
+logger = logging.getLogger("quilt-lambda-s3hash")
+logger.setLevel(os.environ.get("QUILT_LOG_LEVEL", "WARNING"))
+
+DEFAULT_CONCURRENCY = int(os.environ["MPU_CONCURRENCY"])
+MULTIPART_CHECKSUMS = os.environ["MULTIPART_CHECKSUMS"] == "true"
+SERVICE_BUCKET = os.environ["SERVICE_BUCKET"]
+
+SCRATCH_KEY_SERVICE = "user-requests/checksum-upload-tmp"
+SCRATCH_KEY_PER_BUCKET = ".quilt/.checksum-upload-tmp"
+
+# How much seconds before lambda is supposed to timeout we give up.
+SECONDS_TO_CLEANUP = 1
+
+S3: contextvars.ContextVar[S3Client] = contextvars.ContextVar("s3")
+
+
+# Isolated for test-ability.
+def get_user_boto_session(
+    credentials: AWSCredentials,
+) -> aiobotocore.session.AioSession:
+    s = aiobotocore.session.get_session()
+    s.set_credentials(credentials.key, credentials.secret, credentials.token)
+    return s
+
+
+@contextlib.asynccontextmanager
+async def aio_context(credentials: AWSCredentials, concurrency: pydantic.PositiveInt):
+    session = get_user_boto_session(credentials)
+    config = aiobotocore.config.AioConfig(max_pool_connections=concurrency)
+
+    async with session.create_client("s3", config=config) as s3:
+        s3_token = S3.set(s3)
+        try:
+            yield
+        finally:
+            S3.reset(s3_token)
+
+
+class S3hashException(Exception):
```

Comment:
Maybe extract base class for this and PkgpushException to shared? Handling boto errors also makes sense here.

Reply:
ok i don't mind

Reply:
i'll do it

Reply:
done